crates/driver/src/domain/competition/sorting.rs

```diff
@@ -0,0 +1,95 @@
+use {
+    crate::{
+        domain::{
+            competition::{auction::Tokens, order},
+            eth,
+        },
+        util,
+    },
+    chrono::{Duration, Utc},
+    std::sync::Arc,
+};
+
+#[derive(PartialEq, Eq, PartialOrd, Ord)]
+pub enum SortingKey {
+    Int(i32),
+    BigRational(num::BigRational),
+    Timestamp(Option<util::Timestamp>),
+    Bool(bool),
+}
+
+pub trait SortingStrategy: Send + Sync {
+    fn key(&self, order: &order::Order, tokens: &Tokens, solver: &eth::H160) -> SortingKey;
+}
+
+/// Prioritize orders by their class: market orders -> limit orders ->
+/// liquidity.
+///
+/// Market orders are preferred over limit orders, as the expectation is that
+/// they should be immediately fulfillable. Liquidity orders come last, as they
+/// are the most niche and rarely used.
+pub struct OrderClass;
+impl SortingStrategy for OrderClass {
+    fn key(&self, order: &order::Order, _tokens: &Tokens, _solver: &eth::H160) -> SortingKey {
+        SortingKey::Int(match order.kind {
+            order::Kind::Market => 2,
+            order::Kind::Limit { .. } => 1,
+            order::Kind::Liquidity => 0,
+        })
+    }
+}
+
+/// Orders are sorted by their likelihood of being fulfilled, with the most
+/// likely orders coming first. See more details in the `likelihood` function
+/// docs.
+pub struct ExternalPrice;
+impl SortingStrategy for ExternalPrice {
+    fn key(&self, order: &order::Order, tokens: &Tokens, _solver: &eth::H160) -> SortingKey {
+        SortingKey::BigRational(order.likelihood(tokens))
+    }
+}
+
+/// Orders are sorted by their creation timestamp, with the most recent orders
+/// coming first. If `max_order_age` is set, only orders created within the
+/// specified duration will be considered.
+pub struct CreationTimestamp {
+    pub max_order_age: Option<Duration>,
+}
+impl SortingStrategy for CreationTimestamp {
+    fn key(&self, order: &order::Order, _tokens: &Tokens, _solver: &eth::H160) -> SortingKey {
+        SortingKey::Timestamp(match self.max_order_age {
+            Some(max_order_age) => {
+                let earliest_allowed_creation =
+                    u32::try_from((Utc::now() - max_order_age).timestamp()).unwrap_or(u32::MAX);
+                (order.created.0 > earliest_allowed_creation).then_some(order.created)
+            }
+            None => Some(order.created),
+        })
+    }
+}
+
+/// Prioritize orders based on whether the current solver provided the winning
+/// quote for the order.
+pub struct OwnQuotes;
+impl SortingStrategy for OwnQuotes {
```

Comment:
Maybe this one should also have max order age?

Reply:
Yeah, probably makes sense. Disabled by default since it is already defined in the CreationTimestamp strategy.